File: gacha.py
# -*- coding: utf-8 -*-
"""抽卡引擎：按弹（扩展系列）构建稀有度卡池，按各弹发售规格的划档概率抽包。"""
import json
import random
from pathlib import Path

import config

ROOT = Path(__file__).parent
CARDS_DIR = ROOT / "data" / "cards"


class SetDataError(RuntimeError):
    pass


def load_cards(set_id: str) -> list:
    path = CARDS_DIR / f"{set_id}.json"
    if not path.exists():
        raise SetDataError(f"弹 {set_id} 的卡牌数据不存在，请先运行 python fetch_data.py")
    return json.loads(path.read_text(encoding="utf-8"))


def build_pools(cards: list) -> dict:
    """按稀有度分池；符号稀有度（宝石包 ●◆★ 等）先映射为标准稀有度。

    卡面展示仍保留原始记号，映射仅用于概率归类。
    """
    pools = {}
    for c in cards:
        r = c.get("rarity") or "N"
        pools.setdefault(config.RARITY_ALIAS.get(r, r), []).append(c)
    return pools


def _normalize(weights: dict, available: set) -> dict:
    """仅保留池中存在的稀有度并归一化；若全部缺失则返回空表。"""
    w = {r: v for r, v in weights.items() if r in available and v > 0}
    if not w:
        return {}
    total = sum(w.values())
    return {r: v / total for r, v in w.items()}


def _pick_rarity(probs: dict) -> str:
    roll = random.random()
    acc = 0.0
    for r, p in probs.items():
        acc += p
        if roll <= acc:
            return r
    return next(reversed(probs))


def _card_key(card: dict) -> str:
    return f"{card.get('setCode')}__{card.get('cardIndex')}"


def _pick_card(pools: dict, rarity: str | None, cards: list, used: set | None = None) -> dict:
    """按稀有度池均匀抽卡；包内已出现的卡不再出现（同稀有度池去重，池耗尽退回整弹去重）。"""
    if rarity is None:
        # 兜底：整弹均匀抽（特典/礼盒类弹的权重稀有度全部缺失时）
        src = cards
    else:
        src = pools.get(rarity) or cards
    if used:
        fresh = [c for c in src if _card_key(c) not in used]
        if not fresh:
            fresh = [c for c in cards if _card_key(c) not in used]
        if fresh:
            src = fresh
    return dict(random.choice(src))


def resolve_spec(set_id: str, spec_id: str | None) -> dict:
    """按 key 或 id 解析某弹的包规格。"""
    specs = config.set_specs(set_id)
    if spec_id:
        for s in specs:
            if s["key"] == spec_id or s["id"] == spec_id:
                return s
        raise SetDataError(f"弹 {set_id} 不存在规格 {spec_id}")
    return specs[0]


def _slot_probs(slot: dict, set_code: str, spec_id: str, pools: dict):
    """计算某槽位的实际概率表（已按本弹稀有度归一化）。

    权重稀有度在本弹全部缺失时返回 None，表示该槽位退化为整弹均匀抽。
    """
    overrides = config.SET_SPEC_OVERRIDES.get(set_code, {})
    weights = overrides.get((spec_id, slot["name"])) or slot["weights"]
    return _normalize(weights, set(pools)) or None


def _draw_by_slots(slots: list, set_code: str, spec_id: str, pools: dict, cards: list) -> list:
    pack: list = []
    used: set = set()  # 本包已出现的卡（setCode__cardIndex），包内不重复
    for slot in slots:
        probs = _slot_probs(slot, set_code, spec_id, pools)
        if probs:
            card = _pick_card(pools, _pick_rarity(probs), cards, used)
        else:
            card = _pick_card(pools, None, cards, used)
        used.add(_card_key(card))
        card["slotName"] = slot["name"]
        card["slotKind"] = slot["kind"]
        pack.append(card)
    return pack


def draw_pack(set_id: str, spec_id: str | None = None, packs: int = 1) -> list:
    """开包：按弹的发售规格抽 packs 包。

    返回 [pack, ...]，pack 为 [{card, slotName, slotKind}, ...]。
    规格含多个封入变体时（如太晶盛聚 7+3/6+4），每包随机落位。
    """
    cards = load_cards(set_id)
    if not cards:
        raise SetDataError(f"弹 {set_id} 暂无卡牌数据")
    pools = build_pools(cards)
    spec = resolve_spec(set_id, spec_id)
    set_code = set_id.split("__")[0]

    variants = spec.get("variants") or [{"note": spec.get("note", ""), "slots": spec["slots"]}]
    result = []
    for _ in range(packs):
        v = random.choice(variants)
        pack = _draw_by_slots(v["slots"], set_code, spec["id"], pools, cards)
        result.append(pack)
    return result


def spec_probabilities(set_id: str, spec: dict) -> dict:
    """某弹某规格的概率表，含封入变体（供「概率公示」展示）。"""
    cards = load_cards(set_id)
    pools = build_pools(cards)
    set_code = set_id.split("__")[0]
    spec_id = spec["id"]

    def slot_table(slots: list) -> list:
        out = []
        for slot in slots:
            probs = _slot_probs(slot, set_code, spec_id, pools)
            out.append({
                "name": slot["name"],
                "kind": slot["kind"],
                "fallback": probs is None,
                "probabilities": [
                    {"rarity": r, "p": p, "pool": len(pools.get(r, []))}
                    for r, p in sorted((probs or {}).items(), key=lambda kv: -kv[1])
                ],
            })
        return out

    variants = spec.get("variants") or [{"note": spec.get("note", ""), "slots": spec.get("slots", [])}]
    first = variants[0]
    return {
        "id": spec_id,
        "label": spec["label"],
        "note": spec.get("note", ""),
        "price": spec.get("price"),
        "packSize": len(first.get("slots") or spec.get("slots") or []),
        "variants": [
            {"note": v["note"], "slots": slot_table(v["slots"])} for v in variants
        ],
    }


def set_probabilities(set_id: str) -> list:
    """某弹全部规格的概率表。"""
    return [spec_probabilities(set_id, spec) for spec in config.set_specs(set_id)]

File: miniprogram/utils/gacha.js
/* PTCG 简中拆卡引擎 —— 唯一权威实现。
 *
 * 浏览器（exe / 安卓 APK）/ Node / 微信小程序通用：
 *   - 浏览器：window.PTCGGacha
 *   - Node / 小程序：module.exports
 *
 * 约定：规格（spec）由 config.py 导出的元数据提供（含 slots / variants / weights），
 *       本引擎不内嵌任何弹配置，保证 Python 与 JS 概率行为同源。
 *       所有随机入口都接受可选 rng（返回 [0,1) 浮点的函数），默认 Math.random；
 *       注入同种子 rng 即可做 Python↔JS 逐卡对拍（见 tests/）。
 */
"use strict";

(function (root, factory) {
  if (typeof module === "object" && typeof module.exports === "object") {
    module.exports = factory();
  } else {
    root.PTCGGacha = factory();
  }
})(typeof self !== "undefined" ? self : this, function () {
  /* 宝石包等商品的符号稀有度 → 标准稀有度（仅用于卡池归类与概率计算） */
  const RARITY_ALIAS = {
    "●": "C", "○": "C",
    "◆": "U", "◇": "U",
    "★": "R", "★★": "RR", "★★★": "SAR",
  };

  /* 可复现随机源（与 tests/mulberry.py 实现逐位一致，用于跨引擎对拍） */
  function mulberry32(seed) {
    let a = seed >>> 0;
    return function () {
      a = (a + 0x6D2B79F5) >>> 0;
      let t = a;
      t = Math.imul(t ^ (t >>> 15), t | 1);
      t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
      return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
  }

  /* 按稀有度分池；符号稀有度先映射为标准稀有度（卡面展示仍保留原始记号） */
  function buildPools(cards) {
    const pools = {};
    for (const c of cards) {
      const r = c.rarity || "N";
      const k = RARITY_ALIAS[r] || r;
      (pools[k] = pools[k] || []).push(c);
    }
    return pools;
  }

  /* 仅保留池中存在的稀有度并归一化；全部缺失返回 null（槽位退化为整弹均匀抽） */
  function normalizeWeights(weights, available) {
    const w = Object.entries(weights).filter(([r, v]) => available.includes(r) && v > 0);
    if (!w.length) return null;
    const total = w.reduce((a, [, v]) => a + v, 0);
    return Object.fromEntries(w.map(([r, v]) => [r, v / total]));
  }

  function pickRarity(probs, rng) {
    let roll = (rng || Math.random)(), acc = 0;
    for (const [r, p] of Object.entries(probs)) { acc += p; if (roll <= acc) return r; }
    return Object.keys(probs).pop();
  }

  function cardKey(c) { return `${c.setCode}__${c.cardIndex}`; }

  /* 按稀有度池均匀抽卡；包内已出现的卡不再出现（同稀有度池去重，池耗尽退回整弹去重） */
  function pickCard(pools, rarity, cards, used, rng) {
    const rand = rng || Math.random;
    let src = rarity ? (pools[rarity] && pools[rarity].length ? pools[rarity] : cards) : cards;
    if (used) {
      let fresh = src.filter((c) => !used.has(cardKey(c)));
      if (!fresh.length) fresh = cards.filter((c) => !used.has(cardKey(c)));
      if (fresh.length) src = fresh;
    }
    return { ...src[Math.floor(rand() * src.length)] };
  }

  /* 开一包：规格含多个封入变体时（如太晶盛聚 7+3/6+4），每包随机落位 */
  function drawPack(cards, pools, spec, rng) {
    const variants = spec.variants || [{ note: spec.note || "", slots: spec.slots }];
    const v = variants[Math.floor((rng || Math.random)() * variants.length)];
    const available = Object.keys(pools);
    const used = new Set();
    return v.slots.map((slot) => {
      const probs = normalizeWeights(slot.weights, available);
      const card = pickCard(pools, probs ? pickRarity(probs, rng) : null, cards, used, rng);
      used.add(cardKey(card));
      card.slotName = slot.name;
      card.slotKind = slot.kind;
      return card;
    });
  }

  /* 开 packs 包（pools 未传时内部构建） */
  function drawPacks(cards, spec, packs, rng) {
    const pools = buildPools(cards);
    const out = [];
    for (let i = 0; i < packs; i++) out.push(drawPack(cards, pools, spec, rng));
    return out;
  }

  /* 某规格的概率表（含封入变体，供「概率公示」展示） */
  function specProbabilities(spec, pools) {
    const slotTable = (slots) => slots.map((slot) => {
      const probs = normalizeWeights(slot.weights, Object.keys(pools)) || {};
      return {
        name: slot.name, kind: slot.kind, fallback: !probs,
        probabilities: Object.entries(probs)
          .map(([r, p]) => ({ rarity: r, p, pool: (pools[r] || []).length }))
          .sort((a, b) => b.p - a.p),
      };
    });
    const variants = spec.variants || [{ note: spec.note || "", slots: spec.slots }];
    return {
      id: spec.id, label: spec.label, note: spec.note, price: spec.price,
      packSize: variants[0].slots.length,
      variants: variants.map((v) => ({ note: v.note, slots: slotTable(v.slots) })),
    };
  }

  /* 某规格每个稀有度的期望成本（目标卡计算器）：
   * pPack  = 单包至少出 1 张该稀有度的概率（变体按等概率平均）
   * anyPacks = 抽到任意一张该稀有度的期望包数 = 1 / pPack
   * cardPacks = 抽到「指定某一张卡」的期望包数 ≈ anyPacks × 该稀有度池大小
   * 包内去重对期望的影响在包远小于池时忽略不计。 */
  function expectedCost(spec, pools) {
    const variants = spec.variants || [{ note: spec.note || "", slots: spec.slots }];
    const available = Object.keys(pools);
    const total = Object.values(pools).reduce((a, l) => a + l.length, 0);
    return Object.keys(pools).map((r) => {
      let hit = 0;
      for (const v of variants) {
        let miss = 1;
        for (const slot of v.slots) {
          const probs = normalizeWeights(slot.weights, available);
          // 兜底槽位（权重稀有度全部缺失）退化为整弹均匀抽
          const p = probs ? (probs[r] || 0) : total ? ((pools[r] || []).length / total) : 0;
          miss *= 1 - p;
        }
        hit += 1 - miss;
      }
      const pPack = hit / variants.length;
      const anyPacks = pPack > 0 ? 1 / pPack : Infinity;
      const pool = (pools[r] || []).length;
      return {
        rarity: r,
        pPack, pool,
        anyPacks,
        cardPacks: pool ? anyPacks * pool : Infinity,
      };
    }).sort((a, b) => b.pPack - a.pPack);
  }

  return {
    RARITY_ALIAS, mulberry32,
    buildPools, normalizeWeights, pickRarity, pickCard,
    drawPack, drawPacks, specProbabilities, expectedCost,
  };
});
